src/platform/posix/posix_epdesc.c

```diff
@@ -103,7 +103,7 @@ nni_posix_epdesc_doconnect(nni_posix_epdesc *ed)
 		case 0:
 			// Success!
 			nni_posix_epdesc_finish(aio, 0, ed->node.fd);
-			ed->node.fd = -1;
+			// ed->node.fd = -1;
```

Comment:
i'm not sure this is an appropriate change, but without it I was seeing the httpclient test fail during cleanup - because the fd was set to `-1` but not closed, epdesc cleanup code doesn't close it (it's already `-1`) but the poller would deliver a callback after the epdesc had been cleaned up, causing an assert when accessing its mutex.

Please let me know if there's a better way to address this.

Relatedly, in `nni_posix_epdesc_finish()` I don't immediately see where the `nni_posix_pipedesc` that can be allocated there gets free'd - may be that i'm just not familiar with the ownership system there, but could possibly be leaking?

Reply:
I think I may want to refactor some of this somewhat.  Testing with helgrind is actually highlighting that the way we are tweaking the node.fd is actually a potential race.  

So with respect to the pipedesc, if the pipedesc creation succeeds, then rv will be zero, and we will pass that out to the caller in the nni_aio_finish_pipe() logic.  The aio callback then has the responsibility to "own" the pipe, and generally this causes the pipe to get associated with the socket.  That logic is handled in the upper core code, endpt.c and pipe.c.  I don't believe there are any leaks here, and valgrind at least agrees with me in that regard.